File: shared/clients/python/nexus_core_client.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import httpx


@dataclass
class NexusCoreClient:
    base_url: str
    api_key: str
    scopes: list[str]
    actor: str = "operator/engine"
    timeout_seconds: float = 5.0

    def _headers(self, scopes: list[str] | None = None) -> dict[str, str]:
        resolved_scopes = scopes or self.scopes
        headers = {
            "X-NEXUS-GATEWAY-KEY": self.api_key,
            "X-NEXUS-ACTOR": self.actor,
            "X-NEXUS-SCOPES": ",".join(resolved_scopes),
            "Content-Type": "application/json",
        }
        return headers

    def _client(self) -> httpx.Client:
        return httpx.Client(base_url=self.base_url.rstrip("/"), timeout=self.timeout_seconds)

    def list_events(self, cursor: int, limit: int = 100) -> dict[str, Any]:
        with self._client() as client:
            r = client.get(
                f"/v1/events?cursor={cursor}&limit={limit}",
                headers=self._headers(["audit:read", "admin:console:read"]),
            )
            r.raise_for_status()
            return r.json()

    def apply_action(self, payload: dict[str, Any]) -> dict[str, Any]:
        with self._client() as client:
            r = client.post(
                "/v1/actions/apply",
                headers=self._headers(["admin:console:write"]),
                json=payload,
            )
            r.raise_for_status()
            return r.json()

    def create_incident(self, payload: dict[str, Any]) -> dict[str, Any]:
        with self._client() as client:
            r = client.post(
                "/v1/incidents",
                headers=self._headers(["admin:console:write"]),
                json=payload,
            )
            r.raise_for_status()
            return r.json()

    def create_policy_proposal(self, payload: dict[str, Any]) -> dict[str, Any]:
        with self._client() as client:
            r = client.post(
                "/v1/policy-proposals",
                headers=self._headers(["admin:console:write"]),
                json=payload,
            )
            r.raise_for_status()
            return r.json()

File: shared/clients/ts/nexus_core_client.ts
export type NexusCoreClientConfig = {
  baseUrl: string;
  apiKey: string;
  scopes?: string[];
  actor?: string;
};

export class NexusCoreClient {
  constructor(private readonly cfg: NexusCoreClientConfig) {}

  private headers(extra?: Record<string, string>): HeadersInit {
    const h: Record<string, string> = {
      "Content-Type": "application/json",
      "X-NEXUS-CORE-KEY": this.cfg.apiKey,
    };
    if (this.cfg.scopes && this.cfg.scopes.length > 0) {
      h["X-NEXUS-SCOPES"] = this.cfg.scopes.join(",");
    }
    if (this.cfg.actor) {
      h["X-NEXUS-ACTOR"] = this.cfg.actor;
    }
    return { ...h, ...(extra || {}) };
  }

  async listEvents(cursor = 0, limit = 100) {
    const url = `${this.cfg.baseUrl}/v1/events?cursor=${cursor}&limit=${limit}`;
    const resp = await fetch(url, { headers: this.headers() });
    if (!resp.ok) throw new Error(`listEvents failed: ${resp.status}`);
    return resp.json();
  }

  async applyAction(payload: unknown) {
    const resp = await fetch(`${this.cfg.baseUrl}/v1/actions/apply`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify(payload),
    });
    if (!resp.ok) throw new Error(`applyAction failed: ${resp.status}`);
    return resp.json();
  }

  async listIncidents() {
    const resp = await fetch(`${this.cfg.baseUrl}/v1/incidents`, { headers: this.headers() });
    if (!resp.ok) throw new Error(`listIncidents failed: ${resp.status}`);
    return resp.json();
  }

  async listPolicyProposals() {
    const resp = await fetch(`${this.cfg.baseUrl}/v1/policy-proposals`, { headers: this.headers() });
    if (!resp.ok) throw new Error(`listPolicyProposals failed: ${resp.status}`);
    return resp.json();
  }
}
